Add Streams article

tomayac opened the pull request:
Fixes #2703

Changes proposed in this pull request:

- Add the article

[Staged preview](https://deploy-preview-4378--web-dev-staging.netlify.app/streams/)

tomayac commented:
@jpmedley, @jakearchibald, @beaufortfrancois, @samdutton This is a pretty big pull request. I wonder if you have time to review it? If either of you prefers Google Docs over GitHub for reviewing, you can use [this link](https://docs.google.com/document/d/1-2Hbn8uZC3ZWUH4UXTdXyQEvgxOVJCWmEHoso7KocDc/edit?usp=sharing&resourcekey=0-UTxUmY6V6lCc7uPfLioDMw).

jakearchibald commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+  },
+});
+
+let result = '';
+const reader = stream.getReader();
+
+const readStream = async () => {
+  // The `read()` method returns a promise that
+  // resolves when a value has been received.
+  const { done, value } = await reader.read();
+  // Result objects contain two properties:
+  // `done`  - `true` if the stream has already given you all its data.
+  // `value` - Some data. Always `undefined` when `done` is `true`.
+  if (done) {
+    return console.log(`Stream complete.\n${result}`);
+  }
+  result += value;
+  console.log(`Read ${result.length} characters so far`);
+  console.log(`Most recently read chunk: ${value}`);
+  // Read some more, and call this function again.
+  return readStream();
+};
+readStream();
```
I think a recursive function is harder to understand than a loop. Eg:

```suggestion
async function concatStringStream(stream) {
  let result = '';
  const reader = stream.getReader();
  while (true) {
    // The `read()` method returns a promise that
    // resolves when a value has been received.
    const { done, value } = await reader.read();
    // Result objects contain two properties:
    // `done`  - `true` if the stream has already given you all its data.
    // `value` - Some data. Always `undefined` when `done` is `true`.
    if (done) return result;
    result += value;
    console.log(`Read ${result.length} characters so far`);
    console.log(`Most recently read chunk: ${value}`);
  }
}

concatStringStream(stream)
  .then((result) => console.log('Stream complete', result));
```

tomayac replied:
Went for your suggestion.

beaufortfrancois commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+  },
+});
+
+(async () => {
```
Same as before, let's remove this line.

tomayac replied:
OK.

jakearchibald commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+  });
+  // Return the new stream with the headers of the original response.
+  return await { stream: responseStream, headers: response.headers };
+};
```
This example would be much simpler as a transform stream, and also using the `TextDecoderStream` etc. If you're deliberately avoiding those, probably best to cover why.

tomayac replied:
Here, the reader doesn't know about them yet :-) They are being introduced later. I will add a note that states this.

beaufortfrancois commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+  const start = Date.now();
+  for (const char of 'abcd') {
+    // Wait to add write queue.
+    await writer.ready;
```
Was this example to show the use of `ready` or could it have been shorter with something like below?
```javascript
  for (const char of 'abcd') {
    await writer.write(char);
  }
  await writer.close();
```

tomayac replied:
The idea was to show `.ready` indeed.

beaufortfrancois commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+
+A readable stream can be piped to a writable stream through the readable stream's
+[`pipeTo()`](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream/pipeTo) method.
+`ReadableStream.pipeTo()` pipes the current `ReadableStream`to a given`WritableStream` and returns a
```
Typo: space after "given" missing.

tomayac replied:
Fixed.

beaufortfrancois commented on src/site/content/en/blog/streams/index.md:
````diff
@@ -0,0 +1,966 @@
+
+```js
+const readableStream = new ReadableStream({
+  async start(controller) {
````
Nit: Remove unnecessary `async`

tomayac replied:
Done.

beaufortfrancois commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+    controller.enqueue('b');
+    controller.enqueue('c');
+  },
+  async pull(controller) {
```
Nit: Remove unnecessary `async`

tomayac replied:
Done.

beaufortfrancois commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+    controller.enqueue('d');
+    controller.close();
+  },
+  async cancel(reason) {
```
Nit: Remove unnecessary `async`

tomayac replied:
Done.

beaufortfrancois commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+});
+
+const writableStream = new WritableStream({
+  async start(controller) {
```
Nit: Remove unnecessary `async`

tomayac replied:
Done.

beaufortfrancois commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+    // Called by constructor
+    console.log('[start writable]');
+  },
+  async write(chunk, controller) {
```
Nit: Remove unnecessary `async`

tomayac replied:
Done.

beaufortfrancois commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+    // Called when writer.write()
+    console.log('[write]', chunk);
+  },
+  async close(controller) {
```
Nit: Remove unnecessary `async`

tomayac replied:
Done.

beaufortfrancois commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+  async close(controller) {
+    console.log('[close]');
+  },
+  async abort(reason) {
```
Nit: Remove unnecessary `async`

tomayac replied:
Done.

beaufortfrancois commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+  },
+});
+
+(async () => {
```
Once again, it's not needed there.
A comment though would be appreciated such as the console output.

beaufortfrancois replied:
I didn't expect `pull` for instance. Explaining it would be beneficial to users.
```
 [start readable]
 [start writable]
 [write] a
 [write] b
 [pull]
 [write] c
 [write] d
 [close]
 [finished]
```

jakearchibald commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+
+  // Read streamB in a loop.
+  const readerB = streamB.getReader();
+  for (let result = await readerB.read(); !result.done; result = await readerB.read()) {
+    console.log('[B]', result);
+  }
```
```suggestion
  while (true) {
    const result = await readerB.read();
    if (result.done) break;
    console.log('[B]', result);
  }
```

IMO easier to read and avoids duplicating `await readerB.read()`

beaufortfrancois reviewed (changes requested):
I've stopped at "Creating a transform stream" section to breathe a bit ;)
Thanks for creating this content!

jakearchibald commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+
+let startingAB = new ArrayBuffer(1_024);
+const buffer = await readInto(startingAB);
+console.log('The first 1,024 bytes: ', buffer);
```
(unless there are less than 1024 bytes in the stream)

tomayac replied:
Correct :-) Updated the log.

jakearchibald commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+async function readInto(buffer) {
+  let offset = 0;
+
+  while (offset < buffer.byteLength) {
```
Ah, so it isn't required to fill the view you give it? I thought it might fill the view aside from the final chunk.

tomayac replied:
I don't know the answer. This example comes from the explainer.

jakearchibald commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+
+When a writer is created and starts writing to a stream (an **active writer**), it is said to be
+**locked** to it. Only one writer can write to a writable stream at one time. If you want another
+writer to start writing to your stream, you typically need to abort it, before you then attach
```
```suggestion
writer to start writing to your stream, you typically need to release it, before you then attach
```

jakearchibald commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+been processed by the underlying sink.
+
+The queue is managed by a **queuing strategy** for the stream. The strategy's **high water mark**
+defines the total number of chunks that can be contained in the internal queue before backpressure
```
HWM is a size not a count

tomayac replied:
Fixed throughout.

jakearchibald commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+
+The queue is managed by a **queuing strategy** for the stream. The strategy's **high water mark**
+defines the total number of chunks that can be contained in the internal queue before backpressure
+is applied and a **chunk size** that indicates the size to use for each chunk, in bytes.
```
Not necessarily bytes

tomayac replied:
Fixed throughout.

jakearchibald commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+[`WritableStreamDefaultController.error()`](https://developer.mozilla.org/en-US/docs/Web/API/WritableStreamDefaultController/error),
+which causes any future interactions with the associated stream to error.
+
+#### The `queuingStrategy`
```
See comments from the previous `queuingStrategy` section

tomayac replied:
Fixed throughout.

jakearchibald commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+  async abort(reason) {
+    console.log('[abort]', reason);
+  },
+});
```
Maybe create a more practical demo? Eg, creating a writable that appends chunks to a DOM element.

tomayac replied:
Alright, will do.

jakearchibald commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+  async abort(reason) {
+    console.log('[abort]', reason);
+  },
+});
```
If you do the 'write to DOM element' example above, you can use it here to make a more visual example

tomayac replied:
Done.

jakearchibald commented on src/site/content/en/blog/streams/index.md:
````diff
@@ -0,0 +1,966 @@
+The following code sample shows a simple transform stream in action.
+
+```js
+const transformStream = new TransformStream({
````
```suggestion
const textEncoderStream = new TransformStream({
```

Also, maybe mention this exists in the platform already?

tomayac replied:
Also mentioned that this exists now.

jakearchibald reviewed (changes requested):
Coming along nicely!

surma commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+  - streams
+---
+
+The Streams API allows you to programmatically access streams of data received over the network and
```
(_cough_ https://github.com/surma/observables-with-streams)

tomayac replied:
Added "or created by whatever means locally" to make this clear.

jakearchibald commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+    },
+  });
+  // Return the new stream with the headers of the original response.
+  return await { stream: responseStream, headers: response.headers };
```
nah https://jakearchibald.com/2017/await-vs-return-vs-return-await/

tomayac replied:
Thanks! Removed.

surma commented on src/site/content/en/blog/streams/index.md:
````diff
@@ -0,0 +1,966 @@
+`done`.
+
+```js
+let interval = null;
+
+const stream = new ReadableStream({
+  start(controller) {
+    interval = setInterval(() => {
+      // Extract just the hh:mm:ss from the date.
+      let string = new Date().toISOString().replace(/.*?T(.*?)\..*?$/, '$1 ');
+      // Add the string to the stream.
+      console.log(`Enqueued ${string}`);
+      controller.enqueue(string);
+    }, 1_000);
+
+    setTimeout(() => {
+      clearInterval(interval);
+      // Close the stream after 10s.
+      controller.close();
+    }, 10_000);
+  },
+
+  cancel() {
+    // This is called if the reader cancels.
+    clearInterval(interval);
+  },
+});
````
(I have grown quite fond of using `this` in underlying sources. So you could use `this.intervalId = setInterval(...)`

tomayac replied:
Outsourced to its own function now.

samdutton commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+
+You create a readable stream by calling its constructor
+[`ReadableStream()`](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream/ReadableStream).
+It has an optional argument called the stream's `underlyingSource`, which represents an object
```
```suggestion
The constructor has an optional parameter `underlyingSource`, which represents an object
```

beaufortfrancois commented:
@tomayac I was wondering how we could make it more visual to explain uppercase streams without a service worker involved. Here's an attempt: https://uppercase-text-transform-stream.glitch.me/

Code is pretty simple as well:

```js
const LOREM_IPSUM =
  "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sic enim censent, oportunitatis esse beate vivere. Non quam nostram quidem, inquit Pomponius iocans; Itaque primos congressus copulationesque et consuetudinum instituendarum voluntates fieri propter voluptatem; Duo Reges: constructio interrete. Id enim ille summum bonum eu)qumi/an et saepe a)qambi/an appellat, id est animum terrore liberum. An eum discere ea mavis, quae cum plane perdidiceriti nihil sciat? Nam quid possumus facere melius? Alia quaedam dicent, credo, magna antiquorum esse peccata, quae ille veri investigandi cupidus nullo modo ferre potuerit. Itaque illa non dico me expetere, sed legere, nec optare, sed sumere, contraria autem non fugere, sed quasi secernere. Quae qui non vident, nihil umquam magnum ac cognitione dignum amaverunt.";

const upperCaseTextTransformStream = new TransformStream({
  transform(chunk, controller) {
    controller.enqueue(chunk.toUpperCase());
  },
  flush(controller) {
    controller.terminate();
  }
});

const loremIpsumstream = new ReadableStream({
  async start(controller) {
    for (const char of LOREM_IPSUM) {
      firstParagraph.textContent += char;
      await new Promise(resolve => setTimeout(resolve, 10));
      controller.enqueue(char);
    }
    controller.close();
  }
});

(async () => {
  const stream = loremIpsumstream.pipeThrough(upperCaseTextTransformStream);
  const reader = stream.getReader();
  while (true) {
    const { value, done } = await reader.read();
    if (done) break;
    secondParagraph.textContent += value;
  }
})();
```

jakearchibald commented:
@beaufortfrancois made a couple of modification to use writable streams and make it more reusable https://reuseable-transform-demo.glitch.me/

```js
onst LOREM_IPSUM =
  "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sic enim censent, oportunitatis esse beate vivere. Non quam nostram quidem, inquit Pomponius iocans; Itaque primos congressus copulationesque et consuetudinum instituendarum voluntates fieri propter voluptatem; Duo Reges: constructio interrete. Id enim ille summum bonum eu)qumi/an et saepe a)qambi/an appellat, id est animum terrore liberum. An eum discere ea mavis, quae cum plane perdidiceriti nihil sciat? Nam quid possumus facere melius? Alia quaedam dicent, credo, magna antiquorum esse peccata, quae ille veri investigandi cupidus nullo modo ferre potuerit. Itaque illa non dico me expetere, sed legere, nec optare, sed sumere, contraria autem non fugere, sed quasi secernere. Quae qui non vident, nihil umquam magnum ac cognitione dignum amaverunt.";

const wait = (ms) => new Promise(r => setTimeout(r, ms));

function delayedChunksStream(input, { minDelay = 100 } = {}) {
  const chunks = input.slice();
  let delay = wait(minDelay);
  
  
  return new ReadableStream({
    async pull(controller) {
      await delay;
      if (chunks.length === 0) {
        controller.close();
        return;
      }
      controller.enqueue(chunks.shift());
      delay = wait(minDelay);
    }
  });
}

function upperCaseStream() {
  return new TransformStream({
    transform(chunk, controller) {
      controller.enqueue(chunk.toUpperCase());
    },
  });
}

function appendToDomStream(el) {
  return new WritableStream({
    write(chunk) {
      el.append(chunk);
    }
  });
}

const firstP = document.querySelector('.first-p');
const secondP = document.querySelector('.second-p');

const [stream1, stream2] = delayedChunksStream(LOREM_IPSUM.split(/\b/)).tee();

stream1.pipeTo(appendToDomStream(firstP));
stream2.pipeThrough(upperCaseStream()).pipeTo(appendToDomStream(secondP));
```

The above could be reused with the source coming from fetch:

```js
fetch(url).then((response) =>
  response.body
    .pipeThrough(new TextDecoderStream())
    .pipeThrough(upperCaseStream())
    .pipeTo(appendToDomStream(secondP));
);
```

tomayac commented:
@jakearchibald, @beaufortfrancois, @samdutton *Thanks* `<3` for all your reviews and comments. I'm busy with an unrelated PR before that, but will get back to this one by end of week.

surma commented:
Feel free to 100% ignore this, but I did write a [stream visualizer](https://surma.dev/lab/whatwg-stream-visualizer/lab.html) a while ago. If useful, feel free to link/copy.

jpmedley commented:
@tomayac I can't review this until the issues from Jake and François are addressed.

jpmedley commented:
@tomayac Does this cover [byte streams](https://www.chromestatus.com/features/4535319661641728)?

jakearchibald commented:
@jpmedley it does

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+  - streams
+---
+
+The Streams API allows you to programmatically access streams of data received over the network and
```
```suggestion
The Streams API allows you to programmatically access streams of data received over the network
or created by whatever means locally and
```

jakearchibald replied:
I'm unsure about the wording here. It's kinda like saying "Promises allow you to programmatically access a sequence of bluetooth devices or any other local object".

While it's true that [you can get a promise for a sequence of bluetooth devices](https://webbluetoothcg.github.io/web-bluetooth/#dom-bluetooth-getdevices), it isn't really a direct feature of promises.

It seems more correct to say that the fetch API can be used to get a stream of network data.

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+---
+
+The Streams API allows you to programmatically access streams of data received over the network and
+process them with JavaScript. Streaming involves breaking a resource that you want to receive down
```
```suggestion
process them with JavaScript. Streaming involves breaking a resource that you want to receive, send, or transform down
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+---
+
+The Streams API allows you to programmatically access streams of data received over the network and
+process them with JavaScript. Streaming involves breaking a resource that you want to receive down
```
Added "that you want to receive, send, or transform" explicitly.

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+process them with JavaScript. Streaming involves breaking a resource that you want to receive down
+into small chunks, and then processing these chunks bit by bit. While streaming is something
+browsers do anyway when receiving assets like HTML or videos to be shown on webpages, this
+capability has never been available to JavaScript before.
```
```suggestion
capability has never been available to JavaScript before `fetch` with streams was introduced in 2015.

{% aside %}
  Streaming was possible with `XMLHttpRequest`, but it
  [really was not pretty](https://gist.github.com/igrigorik/5736866).
{% endAside %}
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+process them with JavaScript. Streaming involves breaking a resource that you want to receive down
+into small chunks, and then processing these chunks bit by bit. While streaming is something
+browsers do anyway when receiving assets like HTML or videos to be shown on webpages, this
+capability has never been available to JavaScript before.
```
Added the 2015 time stamp and an aside pointing to XHR streaming.

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+capability has never been available to JavaScript before.
+
+Previously, if you wanted to process a resource of some kind (be it a video, or a text file, etc.),
+you would have to download the entire file, wait for it to be deserialized into a suitable format,
```
Is this user-exposed, though?

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+soon as it is available on the client-side, without needing to generate a buffer, string, or blob.
+This unlocks a number of use cases, some of which I list below:
+
+- **Video effects:** piping a readable video stream through a transform stream that applies effects
```
I don't :-(

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+### Chunks
+
+A chunk is a **single piece of data** that is written to or read from a stream. It can be of any
+type; streams can even contain chunks of different types. A chunk will often not be the most atomic
```
```suggestion
type; streams can even contain chunks of different types. Most of the time, a chunk will not be the most atomic
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+### Chunks
+
+A chunk is a **single piece of data** that is written to or read from a stream. It can be of any
+type; streams can even contain chunks of different types. A chunk will often not be the most atomic
```
Done.

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+A chunk is a **single piece of data** that is written to or read from a stream. It can be of any
+type; streams can even contain chunks of different types. A chunk will often not be the most atomic
+unit of data for a given stream. For example, a byte stream might contain chunks consisting of 16
+KiB `Uint8Array` units, instead of single bytes.
```
We could, but it is really just an example without concrete practical use without context.

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+### Transform streams
+
+A transform stream consists of a **pair of streams**: a writable stream, known as its writable side,
+and a readable stream, known as its readable side. In a manner specific to the transform stream in
```
```suggestion
and a readable stream, known as its readable side.
A real-world metaphor for this would be a
[simultaneous interpreter](https://en.wikipedia.org/wiki/Simultaneous_interpretation)
who on-the-fly translates from one language to another.
In a manner specific to the transform stream in
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+
+### Transform streams
+
+A transform stream consists of a **pair of streams**: a writable stream, known as its writable side,
```
https://en.wikipedia.org/wiki/Simultaneous_interpretation it is :-)

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+
+A readable stream can be teed (named after the shape of an uppercase 'T') using its `tee()` method.
+This will lock the stream, making it no longer directly usable; however, it will create **two new
+streams**, called branches, which can be consumed independently.
```
```suggestion
streams**, called branches, which can be consumed independently.
Teeing also is important because streams cannot be rewound or restarted, more about this later.
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+
+A readable stream can be teed (named after the shape of an uppercase 'T') using its `tee()` method.
+This will lock the stream, making it no longer directly usable; however, it will create **two new
+streams**, called branches, which can be consumed independently.
```
Added with a reference to "later".

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+  <figcaption class="w-figcaption">A pipe chain.</figcaption>
+</figure>
+
+## The mechanics of a readable stream
```
Personally I *tend* to dislike this because it leaves you without a complete code snippet (or duplication with the pieces and then the whole). I think it mostly depends on how long the prose goes until the snippet comes. My idea here was to decouple the mechanics from the concrete code.

tomayac replied:
You're three now asking for this, so I will do it and interleave prose with example code :-)

beaufortfrancois replied:
Looking at https://deploy-preview-4378--web-dev-staging.netlify.app/streams/, I still don't see example codes at the top of the article. Is this intended?

tomayac replied:
The code snippets starts here: https://deploy-preview-4378--web-dev-staging.netlify.app/streams/#creating-a-readable-stream. From there onward, I slowly build the example up.

tomayac replied:
I follow the same pattern in https://deploy-preview-4378--web-dev-staging.netlify.app/streams/#creating-a-writable-stream and https://deploy-preview-4378--web-dev-staging.netlify.app/streams/#creating-a-transform-stream.

beaufortfrancois replied:
I would have liked some JS code earlier but it's a personal preference.

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+types of underlying source:
+
+- **Push sources** constantly push data at you when you have accessed them, and it is up to you to
+  start, pause, or cancel access to the stream. Examples include video streams, Server-Sent Events,
```
Good catch.

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+  start, pause, or cancel access to the stream. Examples include video streams, Server-Sent Events,
+  or WebSockets.
+- **Pull sources** require you to explicitly request data from them once connected to. Examples
+  include file access operations via `fetch()` or `XMLHttpRequest` calls.
```
```suggestion
  include HTTP operations via `fetch()` or `XMLHttpRequest` calls.
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+  start, pause, or cancel access to the stream. Examples include video streams, Server-Sent Events,
+  or WebSockets.
+- **Pull sources** require you to explicitly request data from them once connected to. Examples
+  include file access operations via `fetch()` or `XMLHttpRequest` calls.
```
Yepp, true indeed.

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+- **Pull sources** require you to explicitly request data from them once connected to. Examples
+  include file access operations via `fetch()` or `XMLHttpRequest` calls.
+
+The data is read sequentially in small pieces called **chunks**. A chunk can be a single byte, or it
+can be something larger such as a typed array of a certain size. A single stream can contain chunks
```
```suggestion
The data is read sequentially in small pieces called **chunks**.
Chunks can be any kind of JavaScript object, but there are streams that specialize
in sending byte data as `Uint8Array`s. A chunk can be a single byte, or it
can be something larger such as a typed array of a certain size. A single stream can contain chunks
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+- **Pull sources** require you to explicitly request data from them once connected to. Examples
+  include file access operations via `fetch()` or `XMLHttpRequest` calls.
+
+The data is read sequentially in small pieces called **chunks**. A chunk can be a single byte, or it
+can be something larger such as a typed array of a certain size. A single stream can contain chunks
```
Done.

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+ready to be read. An **internal queue** keeps track of the chunks that have not yet been read.
+
+The queue is managed by a **queuing strategy** for the stream. The strategy's **high water mark**
+defines the total number of chunks that can be contained in the internal queue before backpressure
```
Alright!

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+controller that, as the name suggests, allows you to control the stream.
+
+Only one reader can read a stream at a time; when a reader is created and starts reading a stream
+(that is, becomes an **active reader**), it is **locked** to it. If you want another reader to start
```
```suggestion
(that is, becomes an **active reader**), it is **locked** to it. If you want another reader to take over
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+controller that, as the name suggests, allows you to control the stream.
+
+Only one reader can read a stream at a time; when a reader is created and starts reading a stream
+(that is, becomes an **active reader**), it is **locked** to it. If you want another reader to start
```
Right. Replaced with "take over reading"

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+- `start(controller)`: This method is called immediately when the object is constructed. The
+  contents of this method should aim to get access to the stream source, and do anything else
+  required to set up the stream functionality. If this process is to be done asynchronously, it can
+  return a promise to signal success or failure. The `controller` parameter passed to this method is
```
```suggestion
  return a promise to signal success or failure. A potentially existing `pull` will not be called until this promise
  fulfills. The stream will become errored if it rejects. The `controller` parameter passed to this method is
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+- `start(controller)`: This method is called immediately when the object is constructed. The
+  contents of this method should aim to get access to the stream source, and do anything else
+  required to set up the stream functionality. If this process is to be done asynchronously, it can
+  return a promise to signal success or failure. The `controller` parameter passed to this method is
```
Done.

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+  required to set up the stream functionality. If this process is to be done asynchronously, it can
+  return a promise to signal success or failure. The `controller` parameter passed to this method is
+  a
+  [`ReadableStreamDefaultController`](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStreamDefaultController)
+  by default.
```
```suggestion
  [`ReadableStreamDefaultController`](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStreamDefaultController).
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+  return a promise to signal success or failure. The `controller` parameter passed to this method is
+  a
+  [`ReadableStreamDefaultController`](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStreamDefaultController)
+  by default.
```
Deferring to later then.

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+const stream = new ReadableStream({
+  start(controller) {
+    interval = setInterval(() => {
+      // Extract just the hh:mm:ss from the date.
+      let string = new Date().toISOString().replace(/.*?T(.*?)\..*?$/, '$1 ');
```
```suggestion
      let string = new Date().toLocaleTimeString();
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+  start(controller) {
+    interval = setInterval(() => {
+      // Extract just the hh:mm:ss from the date.
+      let string = new Date().toISOString().replace(/.*?T(.*?)\..*?$/, '$1 ');
```
Right. I was overthinking it. All we need is a simple string.

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+  console.log(`Read ${result.length} characters so far`);
+  console.log(`Most recently read chunk: ${value}`);
+  // Read some more, and call this function again.
+  return readStream();
```
```suggestion
  readStream();
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+  console.log(`Read ${result.length} characters so far`);
+  console.log(`Most recently read chunk: ${value}`);
+  // Read some more, and call this function again.
+  return readStream();
```
I tend to err on the side of redundant `return`s, but have removed it.

tomayac commented on src/site/content/en/blog/streams/index.md:
````diff
@@ -0,0 +1,966 @@
+the whole document to be downloaded, which can make a huge difference when dealing with large files.
+
+```js
+// Install immediately without waiting.
+self.addEventListener('install', (event) => {
+  self.skipWaiting();
+});
+
+// Claim all clients immediately.
+self.addEventListener('activate', (event) => {
+  clients.claim();
+});
+
````
```suggestion
```

tomayac commented on src/site/content/en/blog/streams/index.md:
````diff
@@ -0,0 +1,966 @@
+the whole document to be downloaded, which can make a huge difference when dealing with large files.
+
+```js
+// Install immediately without waiting.
````
Killed.

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+  // After uppercasing, need to encode chunks again.
+  const textEncoder = new TextEncoder();
+  // Fetch the initial document and get access to the
+  // reader of the readable stream.
+  const response = await fetch(url);
```
```suggestion
  // reader of the readable stream. Error handling skipped for brevity.
  const response = await fetch(url);
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+  const textEncoder = new TextEncoder();
+  // Fetch the initial document and get access to the
+  // reader of the readable stream.
+  const response = await fetch(url);
```
"Skipped for brevity" comment added.

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+    },
+
+    // Called when the reader cancels.
+    cancel() {
```
Not in this snippet.

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+    },
+  });
+  // Return the new stream with the headers of the original response.
+  return await { stream: responseStream, headers: response.headers };
```
```suggestion
  return { stream: responseStream, headers: response.headers };
```

tomayac commented on src/site/content/en/blog/streams/index.md:
````diff
@@ -0,0 +1,966 @@
+  // Return the new stream with the headers of the original response.
+  return await { stream: responseStream, headers: response.headers };
+};
+```
````
```suggestion
```

{% Aside %}
  This example would be much simpler as a transform stream,
  which I will introduce further down in the article. Stay tuned!
{% endAside %}
```

tomayac commented on src/site/content/en/blog/streams/index.md:
````diff
@@ -0,0 +1,966 @@
+  return await { stream: responseStream, headers: response.headers };
+};
+```
+
````
```suggestion

{% Aside %}
  This example would be much simpler as a transform stream,
  which I will introduce further down in the article. Stay tuned!
{% endAside %}

```

tomayac commented on src/site/content/en/blog/streams/index.md:
````diff
@@ -0,0 +1,966 @@
+```
+
+{% Aside 'caution' %} Asynchronous iteration is not yet implemented in any browser. {% endAside %}
+
+#### Asynchronous iteration helper function
````
```suggestion
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+
+{% Aside 'caution' %} Asynchronous iteration is not yet implemented in any browser. {% endAside %}
+
+#### Asynchronous iteration helper function
```
Removed.

tomayac commented on src/site/content/en/blog/streams/index.md:
````diff
@@ -0,0 +1,966 @@
+
+```js
+const readableStream = new ReadableStream({
+  async start(controller) {
````
```suggestion
  start(controller) {
```

tomayac commented on src/site/content/en/blog/streams/index.md:
````diff
@@ -0,0 +1,966 @@
+
+```js
+const readableStream = new ReadableStream({
+  async start(controller) {
````
Done.

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+    controller.enqueue('b');
+    controller.enqueue('c');
+  },
+  async pull(controller) {
```
```suggestion
  pull(controller) {
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+    controller.enqueue('b');
+    controller.enqueue('c');
+  },
+  async pull(controller) {
```
Done.

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+    controller.enqueue('d');
+    controller.close();
+  },
+  async cancel(reason) {
```
```suggestion
  cancel(reason) {
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+    controller.enqueue('d');
+    controller.close();
+  },
+  async cancel(reason) {
```
Done.

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+
+let startingAB = new ArrayBuffer(1_024);
+const buffer = await readInto(startingAB);
+console.log('The first 1,024 bytes: ', buffer);
```
```suggestion
console.log('The first 1,024 bytes (or less): ', buffer);
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+  let offset = 0;
+
+  while (offset < buffer.byteLength) {
+    const { value: view, done } = await reader.read(
+      new Uint8Array(buffer, offset, buffer.byteLength - offset),
```
```suggestion
    const view = new Uint8Array(buffer, offset, buffer.byteLength - offset);
    const { value, done } = await reader.read(view);
    buffer = value.buffer;
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+
+  while (offset < buffer.byteLength) {
+    const { value: view, done } = await reader.read(
+      new Uint8Array(buffer, offset, buffer.byteLength - offset),
```
Yes :-)

tomayac commented on src/site/content/en/blog/streams/index.md:
````diff
@@ -0,0 +1,966 @@
+
+```js
+const writableStream = new WritableStream({
+  async start(controller) {
````
```suggestion
  start(controller) {
```

tomayac commented on src/site/content/en/blog/streams/index.md:
````diff
@@ -0,0 +1,966 @@
+
+```js
+const writableStream = new WritableStream({
+  async start(controller) {
````
Done.

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+  async start(controller) {
+    console.log('[start]');
+  },
+  async write(chunk, controller) {
```
```suggestion
  write(chunk, controller) {
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+  async start(controller) {
+    console.log('[start]');
+  },
+  async write(chunk, controller) {
```
Done.

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+  async write(chunk, controller) {
+    console.log('[write]', chunk);
+    // Wait for next write.
+    await new Promise((resolve) => setTimeout(resolve, 1000));
```
```suggestion
    return new Promise((resolve) => setTimeout(resolve, 1000));
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+  async write(chunk, controller) {
+    console.log('[write]', chunk);
+    // Wait for next write.
+    await new Promise((resolve) => setTimeout(resolve, 1000));
```
Sure. Done.

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+    // Wait for next write.
+    await new Promise((resolve) => setTimeout(resolve, 1000));
+  },
+  async close(controller) {
```
```suggestion
  close(controller) {
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+    // Wait for next write.
+    await new Promise((resolve) => setTimeout(resolve, 1000));
+  },
+  async close(controller) {
```
Done.

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+  async close(controller) {
+    console.log('[close]');
+  },
+  async abort(reason) {
```
```suggestion
  abort(reason) {
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+  async close(controller) {
+    console.log('[close]');
+  },
+  async abort(reason) {
```
Done.

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+
+A readable stream can be piped to a writable stream through the readable stream's
+[`pipeTo()`](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream/pipeTo) method.
+`ReadableStream.pipeTo()` pipes the current `ReadableStream`to a given`WritableStream` and returns a
```
```suggestion
`ReadableStream.pipeTo()` pipes the current `ReadableStream`to a given `WritableStream` and returns a
```

tomayac commented on src/site/content/en/blog/streams/index.md:
````diff
@@ -0,0 +1,966 @@
+
+```js
+const readableStream = new ReadableStream({
+  async start(controller) {
````
```suggestion
  start(controller) {
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+    controller.enqueue('d');
+    controller.close();
+  },
+  async cancel(reason) {
```
```suggestion
  cancel(reason) {
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+});
+
+const writableStream = new WritableStream({
+  async start(controller) {
```
```suggestion
  start(controller) {
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+    // Called by constructor
+    console.log('[start writable]');
+  },
+  async write(chunk, controller) {
```
```suggestion
  write(chunk, controller) {
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+    // Called when writer.write()
+    console.log('[write]', chunk);
+  },
+  async close(controller) {
```
```suggestion
  close(controller) {
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+  async close(controller) {
+    console.log('[close]');
+  },
+  async abort(reason) {
```
```suggestion
  abort(reason) {
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+streams**, called branches, which can be consumed independently.
+
+<figure class="w-figure">
+  <img src="./streams.svg" width="2099" height="1127" alt="Diagram of a pipe chain consisting of a readable stream coming from a call to the fetch API that is then piped through a transform stream whose output is teed and then sent to the browser for the first resulting readable stream and to the service worker cache for the second resulting readable stream.">
```
Do you mean `pipeTo`? The data is `pipeThrough()`'ed a transform stream. Not sure about how to add this to the diagram without looking contrived. Do you?

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+  // Create two `ReadableStream`s.
+  const [streamA, streamB] = readableStream.tee();
+
+  // Read streamA iteratively one by one.
```
The idea here is to show different ways of accessing the data. I will make this clearer.

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,974 @@
+  // Create two `ReadableStream`s.
+  const [streamA, streamB] = readableStream.tee();
+
+  // Read streamA iteratively one by one.
```
```suggestion
  // Read streamA iteratively one by one. Typically, you
  // would not do it this way, but you certainly can.
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,974 @@
+  console.log('[A]', await readerA.read()); //=> {value: "d", done: false}
+  console.log('[A]', await readerA.read()); //=> {value: undefined, done: true}
+
+  // Read streamB in a loop.
```
```suggestion
  // Read streamB in a loop. This is the more common way
  // to read data from the stream.
```

tomayac commented on src/site/content/en/blog/streams/index.md:
````diff
@@ -0,0 +1,974 @@
+The code sample below shows all steps in action.
+
+```js
+const writableStream = new WritableStream({
+  start(controller) {
+    console.log('[start]');
+  },
+  write(chunk, controller) {
+    console.log('[write]', chunk);
+    // Wait for next write.
+    return new Promise((resolve) => setTimeout(resolve, 1_000));
+  },
+  close(controller) {
+    console.log('[close]');
+  },
+  abort(reason) {
+    console.log('[abort]', reason);
+  },
+});
+
+(async () => {
+  const writer = writableStream.getWriter();
+  const start = Date.now();
+  for (const char of 'abcd') {
+    // Wait to add write queue.
+    await writer.ready;
+    console.log('[ready]', Date.now() - start, 'ms');
+    // The Promise is resolved after the write finishes.
+    writer.write(char);
+  }
+  await writer.close();
+})();
````
```suggestion
const writableStream = new WritableStream({
  async start(controller) {
    console.log('[start]');
  },
  async write(chunk, controller) {
    console.log('[write]', chunk);
    // Wait for next write.
    await new Promise((resolve) => setTimeout(() => {
      document.body.textContent += chunk;
      resolve();
    }, 1_000));
  },
  async close(controller) {
    console.log('[close]');
  },
  async abort(reason) {
    console.log('[abort]', reason);
  },
});

const writer = writableStream.getWriter();
const start = Date.now();
for (const char of 'abcdefghijklmnopqrstuvwxyz') {
  // Wait to add write queue.
  await writer.ready;
  console.log('[ready]', Date.now() - start, 'ms');
  // The Promise is resolved after the write finishes.
  writer.write(char);
}
await writer.close();
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,977 @@
+  },
+});
+
+(async () => {
+  // Create two `ReadableStream`s.
+  const [streamA, streamB] = readableStream.tee();
+
+  // Read streamA iteratively one by one. Typically, you
+  // would not do it this way, but you certainly can.
+  const readerA = streamA.getReader();
+  console.log('[A]', await readerA.read()); //=> {value: "a", done: false}
+  console.log('[A]', await readerA.read()); //=> {value: "b", done: false}
+  console.log('[A]', await readerA.read()); //=> {value: "c", done: false}
+  console.log('[A]', await readerA.read()); //=> {value: "d", done: false}
+  console.log('[A]', await readerA.read()); //=> {value: undefined, done: true}
+
+  // Read streamB in a loop. This is the more common way
+  // to read data from the stream.
+  const readerB = streamB.getReader();
+  while (true) {
+    const result = await readerB.read();
+    if (result.done) break;
+    console.log('[B]', result);
+  }
+})();
```
```suggestion
// Create two `ReadableStream`s.
const [streamA, streamB] = readableStream.tee();

// Read streamA iteratively one by one. Typically, you
// would not do it this way, but you certainly can.
const readerA = streamA.getReader();
console.log('[A]', await readerA.read()); //=> {value: "a", done: false}
console.log('[A]', await readerA.read()); //=> {value: "b", done: false}
console.log('[A]', await readerA.read()); //=> {value: "c", done: false}
console.log('[A]', await readerA.read()); //=> {value: "d", done: false}
console.log('[A]', await readerA.read()); //=> {value: undefined, done: true}

// Read streamB in a loop. This is the more common way
// to read data from the stream.
const readerB = streamB.getReader();
while (true) {
  const result = await readerB.read();
  if (result.done) break;
  console.log('[B]', result);
}
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+  },
+});
+
+(async () => {
```
So be it.

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,977 @@
+    controller.enqueue('b');
+    controller.enqueue('c');
+  },
+  async pull(controller) {
```
```suggestion
  pull(controller) {
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,977 @@
+  },
+  write(chunk, controller) {
+    // Called when writer.write()
+    console.log('[write]', chunk);
```
```suggestion
    console.log('[write]', chunk);
    // Wait for next write.
    await new Promise((resolve) => setTimeout(() => {
      document.body.textContent += chunk;
      resolve();
    }, 1_000));
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,977 @@
+  },
+});
+
+(async () => {
+  await readableStream.pipeTo(writableStream);
+  console.log('[finished]');
+})();
```
```suggestion
await readableStream.pipeTo(writableStream);
console.log('[finished]');
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,977 @@
+    console.log('[start writable]');
+  },
+  write(chunk, controller) {
+    // Called when writer.write()
```
```suggestion
    // Called upon writer.write()
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,977 @@
+
+## Acknowledgements
+
+This article was reviewed by [Joe Medley](https://github.com/jpmedley).
```
```suggestion
This article was reviewed by
[Jake Archibald](https://jakearchibald.com/),
[François Beaufort](https://github.com/beaufortfrancois),
[Sam Dutton](https://samdutton.com/), and
[Joe Medley](https://github.com/jpmedley).
```

tomayac commented on src/site/content/en/blog/streams/index.md:
````diff
@@ -0,0 +1,982 @@
+`done`.
+
+```js
+let interval = null;
+
+const stream = new ReadableStream({
+  start(controller) {
+    interval = setInterval(() => {
+      let string = new Date().toLocaleTimeString();
+      // Add the string to the stream.
+      console.log(`Enqueued ${string}`);
+      controller.enqueue(string);
+    }, 1_000);
+
+    setTimeout(() => {
+      clearInterval(interval);
+      // Close the stream after 10s.
+      controller.close();
+    }, 10_000);
````
```suggestion
const startEnqueuing = (controller) => {
  let interval = null;
  interval = setInterval(() => {
    let string = new Date().toLocaleTimeString();
    // Add the string to the stream.
    console.log(`Enqueued ${string}`);
    controller.enqueue(string);
  }, 1_000);

  setTimeout(() => {
    clearInterval(interval);
    // Close the stream after 10s.
    controller.close();
  }, 10_000);
};

const stream = new ReadableStream({
  start(controller) {
    startEnqueuing(controller);
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,982 @@
+  },
+});
+
+let result = '';
+const reader = stream.getReader();
+
+const readStream = async () => {
+  // The `read()` method returns a promise that
+  // resolves when a value has been received.
+  const { done, value } = await reader.read();
+  // Result objects contain two properties:
+  // `done`  - `true` if the stream has already given you all its data.
+  // `value` - Some data. Always `undefined` when `done` is `true`.
+  if (done) {
+    return console.log(`Stream complete.\n${result}`);
+  }
+  result += value;
+  console.log(`Read ${result.length} characters so far`);
+  console.log(`Most recently read chunk: ${value}`);
+  // Read some more, and call this function again.
+  readStream();
+};
+readStream();
```
```suggestion
async function concatStringStream(stream) {
  let result = '';
  const reader = stream.getReader();
  while (true) {
    // The `read()` method returns a promise that
    // resolves when a value has been received.
    const { done, value } = await reader.read();
    // Result objects contain two properties:
    // `done`  - `true` if the stream has already given you all its data.
    // `value` - Some data. Always `undefined` when `done` is `true`.
    if (done) return result;
    result += value;
    console.log(`Read ${result.length} characters so far`);
    console.log(`Most recently read chunk: ${value}`);
  }
}
concatStringStream(stream).then((result) => console.log('Stream complete', result));
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,982 @@
+- [Streams polyfill](https://github.com/MattiasBuelens/web-streams-polyfill)
+- [2016—the year of web streams](https://jakearchibald.com/2016/streams-ftw/)
+- [Async iterators and generators](https://jakearchibald.com/2017/async-iterators-and-generators/)
+
```
```suggestion
- [Stream Visualizer](https://surma.dev/lab/whatwg-stream-visualizer/lab.html)

```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,982 @@
+This article was reviewed by
+[Jake Archibald](https://jakearchibald.com/),
+[François Beaufort](https://github.com/beaufortfrancois),
+[Sam Dutton](https://samdutton.com/), and
```
```suggestion
[Sam Dutton](https://samdutton.com/),
[Surma](https://surma.dev/), and
```

tomayac commented on src/site/content/en/blog/streams/index.md:
````diff
@@ -0,0 +1,983 @@
+The following code sample shows a simple transform stream in action.
+
+```js
+const textEncoderStream = new TransformStream({
````
```suggestion
// Note that `TextEncoderStream` and `TextDecoderStream` exist now.
// This example shows how you would have done it before.
const textEncoderStream = new TransformStream({
```

tomayac commented on src/site/content/en/blog/streams/index.md:
````diff
@@ -0,0 +1,984 @@
+    })(),
+  );
+});
+
+const shoutFetch = async (url) => {
+  // Chunks arrive encoded, so need to decode them.
+  const textDecoder = new TextDecoder('utf-8', { stream: true });
+  // After uppercasing, need to encode chunks again.
+  const textEncoder = new TextEncoder();
+  // Fetch the initial document and get access to the
+  // reader of the readable stream. Error handling skipped for brevity.
+  const response = await fetch(url);
+  const readableStream = response.body;
+  const reader = readableStream.getReader();
+  // Create a new stream.
+  const responseStream = new ReadableStream({
+    // Called whenever the internal queue of chunks is not full.
+    async pull(controller) {
+      // Read the next chunk.
+      let result = await reader.read();
+      // Close the stream when done.
+      if (result.done) {
+        return controller.close();
+      }
+      // Uppercase each chunk (after decoding it).
+      const upperCaseValue = textDecoder.decode(result.value).toUpperCase();
+      // Enqueue the uppercased value (afer re-encoding it).
+      controller.enqueue(textEncoder.encode(upperCaseValue));
+      return result;
+    },
+
+    // Called when the reader cancels.
+    cancel() {
+      reader.cancel();
+    },
+  });
+  // Return the new stream with the headers of the original response.
+  return { stream: responseStream, headers: response.headers };
+};
+```
````
```js
function upperCaseStream() {
  return new TransformStream({
    transform(chunk, controller) {
      controller.enqueue(chunk.toUpperCase());
    },
  });
}

function appendToDomStream(el) {
  return new WritableStream({
    write(chunk) {
      el.append(chunk);
    }
  });
}

fetch('./lorem-psum.txt').then((response) =>
  response.body
    .pipeThrough(new TextDecoderStream())
    .pipeThrough(upperCaseStream())
    .pipeTo(appendToDomStream(document.body));
);
```

tomayac commented on src/site/content/en/blog/streams/index.md:
````diff
@@ -0,0 +1,984 @@
+concatStringStream(stream).then((result) => console.log('Stream complete', result));
+```
+
+The next (a bit contrived) code sample shows how you could implement a "shouting" service worker
+that uppercases all text by consuming the returned `fetch()` response promise
````
```suggestion
The next (a bit contrived) code sample shows how you could implement a "shouting" version of `fetch()`
that uppercases all text by consuming the returned response promise
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,986 @@
+      return new Response(result.stream, { headers: result.headers });
+    })(),
+  );
+});
+
+const shoutFetch = async (url) => {
+  // Chunks arrive encoded, so need to decode them.
+  const textDecoder = new TextDecoder('utf-8', { stream: true });
+  // After uppercasing, need to encode chunks again.
+  const textEncoder = new TextEncoder();
+  // Fetch the initial document and get access to the
+  // reader of the readable stream. Error handling skipped for brevity.
+  const response = await fetch(url);
+  const readableStream = response.body;
+  const reader = readableStream.getReader();
+  // Create a new stream.
+  const responseStream = new ReadableStream({
+    // Called whenever the internal queue of chunks is not full.
+    async pull(controller) {
+      // Read the next chunk.
+      let result = await reader.read();
+      // Close the stream when done.
+      if (result.done) {
+        return controller.close();
+      }
+      // Uppercase each chunk (after decoding it).
+      const upperCaseValue = textDecoder.decode(result.value).toUpperCase();
+      // Enqueue the uppercased value (afer re-encoding it).
+      controller.enqueue(textEncoder.encode(upperCaseValue));
+      return result;
+    },
+
+    // Called when the reader cancels.
+    cancel() {
+      reader.cancel();
+    },
+  });
+  // Return the new stream with the headers of the original response.
+  return { stream: responseStream, headers: response.headers };
+};
```
```suggestion
function upperCaseStream() {
  return new TransformStream({
    transform(chunk, controller) {
      controller.enqueue(chunk.toUpperCase());
    },
  });
}

function appendToDomStream(el) {
  return new WritableStream({
    write(chunk) {
      el.append(chunk);
    }
  });
}

fetch('./lorem-ipsum.txt').then((response) =>
  response.body
    .pipeThrough(new TextDecoderStream())
    .pipeThrough(upperCaseStream())
    .pipeTo(appendToDomStream(document.body));
);
```

jpmedley commented:
@tomayac Is everyone else done?

tomayac commented:
> @tomayac Is everyone else done?

I'm still busy with some of their feedback. Give me a little more time, I will ping here once I'm done. Thanks!

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+
+The queue is managed by a **queuing strategy** for the stream. The strategy's **high water mark**
+defines the total number of chunks that can be contained in the internal queue before backpressure
+is applied and a **chunk size** that indicates the size to use for each chunk, in bytes.
```
Fixed. MDN needs some fixing, too. I'll try once I land this.

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+ready to be read. An **internal queue** keeps track of the chunks that have not yet been read.
+
+The queue is managed by a **queuing strategy** for the stream. The strategy's **high water mark**
+defines the total number of chunks that can be contained in the internal queue before backpressure
```
Fixed. Again MDN needs some cleansing, too.

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+- `highWaterMark`: A non-negative integer that defines the total number of chunks that can be
+  contained in the internal queue before backpressure is applied.
+- `size(chunk)`: A method containing a parameter chunk that indicates the size to use for each
+  chunk, in bytes.
```
Fixed.

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+It is an object that optionally defines a queuing strategy for the stream, which takes two
+parameters:
+
+- `highWaterMark`: A non-negative integer that defines the total number of chunks that can be
```
Fixed.

tomayac commented:
@jpmedley OK, I have addressed the majority of the others' feedback. You can do a round of reviewing now. Thanks for your work 🙏, I know it's a long article.

kaycebasques commented:
Hello! This is an automated review by our custom [reviewbot](https://github.com/kaycebasques/reviewbot). It updates automatically when code or GitHub comments in this pull request are created or updated.

## Requested changes

If there are any common problems with the content files you created or modified, they will be listed here.

### `src/site/content/en/blog/streams/index.md`

* Please break lines at 100 characters maximum. Long lines can be difficult to work with in GitHub's pull request UI. The following tools can help speed up the process: [Rewrap](https://marketplace.visualstudio.com/items?itemName=stkb.rewrap) (VS Code), [Docs To X](https://gsuite.google.com/u/0/marketplace/app/docs_to_x/243920501110) (Google Docs; only available for Googlers). Affected lines: 21, 54, 106, 163, 501, 502, 503, 513, 514, 515, 593, 594, 595
<!-- Comment ID: reviewbot-automated-comment -->

jpmedley commented:
I'll start since this is starting to get a bit old.

@beaufortfrancois @jakearchibald I humbly ask that in the future, you please click approved when your reviews are finished. This gives me assurance that nothing I edit will be rewritten. (I hope you haven't been told you don't need to.)

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,980 @@
+  network and process them as desired.
+authors:
+  - thomassteiner
+date: 2021-01-18
```
```suggestion
date: 2021-02-03
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,980 @@
+  network and process them as desired.
+authors:
+  - thomassteiner
+date: 2021-01-18
```
Done.

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,980 @@
+
+## Core concepts
+
+Before I go into details on the various types of streams, let me introduce some core concepts first.
```
Indeed.

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,980 @@
+### Chunks
+
+A chunk is a **single piece of data** that is written to or read from a stream. It can be of any
+type; streams can even contain chunks of different types. Most of the time, a chunk will not be the most atomic
```
You understand correctly, and it is important to know, since based on this you can't make any assumptions about a chunk and need to be careful when you process them.

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,980 @@
+
+A writable stream represents a destination for data, into which you can write. In other words, data
+**goes in** to a writable stream. Concretely, a writable stream is an instance of the
+`WritableStream` class.
```
😃 This makes me happy.

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,980 @@
+Once a pipe chain is constructed, it will propagate signals regarding how fast chunks should flow
+through it. If any step in the chain cannot yet accept chunks, it propagates a signal backwards
+through the pipe chain, until eventually the original source is told to stop producing chunks so
+fast. This process of **normalizing flow** from the original source according to how fast the chain
+can process chunks is called backpressure.
```
It was indeed somewhat redundant.

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,980 @@
+  by default.
+- `pull(controller)`: Can be used to control the stream as more chunks are fetched. It
+  is called repeatedly as long as the stream's internal queue of chunks is not full, up until the queue
+  reaches its high water mark. If `pull()` returns a promise, it will not be called again until that
+  promise fulfills; if the promise rejects, the stream will become errored.
```
```suggestion
  reaches its high water mark. If the result of calling `pull()` is a promise, `pull()` will not be called again until said
  promise fulfills. If the promise rejects, the stream will become errored.
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,980 @@
+  by default.
+- `pull(controller)`: Can be used to control the stream as more chunks are fetched. It
+  is called repeatedly as long as the stream's internal queue of chunks is not full, up until the queue
+  reaches its high water mark. If `pull()` returns a promise, it will not be called again until that
+  promise fulfills; if the promise rejects, the stream will become errored.
```
I have rephrased this and split it into two sentences.

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+- [`ReadableByteStreamController.enqueue()`](https://developer.mozilla.org/en-US/docs/Web/API/ReadableByteStreamController/enqueue)
+  enqueues a given chunk in the associated stream.
+- [`ReadableByteStreamController.error()`](https://developer.mozilla.org/en-US/docs/Web/API/ReadableByteStreamController/error)
+  causes any future interactions with the associated stream to error.
```
For `ReadableStreamDefaultController()`, this was actually [part of the beta release post](https://blog.chromium.org/2021/01/chrome-89-beta-advanced-hardware.html#:~:text=Expose%20ReadableStreamDefaultController%20interface). It should actually be `ReadableByteStreamDefaultController` here, I missed that.

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,980 @@
+  promise fulfills; if the promise rejects, the stream will become errored.
+- `cancel(reason)`: Called when the stream consumer cancels the stream.
+
+The `ReadableStreamDefaultController` supports the following methods:
```
```suggestion
The `ReadableByteStreamController` supports the following methods:
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,966 @@
+- [`ReadableByteStreamController.error()`](https://developer.mozilla.org/en-US/docs/Web/API/ReadableByteStreamController/error)
+  causes any future interactions with the associated stream to error.
+
+#### The `queuingStrategy`
```
You get what you ask for :-)

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,980 @@
+
+- If a chunk is available, the promise will be fulfilled with an object of the form<br>
+  `{ value: chunk, done: false }`.
+- If the stream becomes closed, the promise will be fulfilled with an object of the form<br>
```
These are actually the terms the spec uses. A stream "becomes" either "errored" or "closed".

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,980 @@
+  `{ value: chunk, done: false }`.
+- If the stream becomes closed, the promise will be fulfilled with an object of the form<br>
+  `{ value: undefined, done: true }`.
+- If the stream becomes errored, the promise will be rejected with the relevant error.
```
Same as before, the spec term is "a stream becomes errored".

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,980 @@
+
+The code sample below shows all the steps in action. You first create a `ReadableStream` that in its
+`underlyingSource` argument defines a `start` method that tells the stream's `controller` to
+`enqueue()` the current time during 10 seconds and then to `close()` the stream. You consume this
```
```suggestion
`enqueue()` a timestamp every second during ten seconds and then to `close()` the stream. You consume this
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,980 @@
+
+The code sample below shows all the steps in action. You first create a `ReadableStream` that in its
+`underlyingSource` argument defines a `start` method that tells the stream's `controller` to
+`enqueue()` the current time during 10 seconds and then to `close()` the stream. You consume this
```
I have rephrased.

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,980 @@
+    // Add the string to the stream.
+    console.log(`Enqueued ${string}`);
+    controller.enqueue(string);
+  }, 1_000);
```
It's called [numeric separators](https://2ality.com/2018/02/numeric-separators.html). Just a way to group numbers in a more readable way. `1_000` is the same as `1000`.

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,980 @@
+
+### Asynchronous iteration
+
+Checking upon each `read()` iteration if the stream is `done` may not be the most convenient API.
```
```suggestion
Checking upon each `read()` loop iteration if the stream is `done` may not be the most convenient API.
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,980 @@
+
+### Asynchronous iteration
+
+Checking upon each `read()` iteration if the stream is `done` may not be the most convenient API.
```
It was missing the word "loop", probably that's why it sounded wrong :-)

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,982 @@
+or [`CountQueuingStrategy`](https://developer.mozilla.org/en-US/docs/Web/API/CountQueuingStrategy)
+for this object value. If no `queuingStrategy` is supplied, the default used is the same as a
+`CountQueuingStrategy` with a `highWaterMark` of `1`. {% endAside %}
+
```
```suggestion
```

tomayac commented:
Alright, reviewers, I have addressed (almost) all of your comments. I think this is getting closer to the finishing line. I've re-requested your eyes once more, thanks in advance!

robdodson commented:
@tomayac is the excalidraw file being used? I tried to do a quick search but couldn't find where it was being loaded.

tomayac commented:
> @tomayac is the excalidraw file being used? I tried to do a quick search but couldn't find where it was being loaded.

The source `.excalidraw` file is not being used, but the exported-from-it `.svg` is. Is there a better place to keep the source around in case we need to make edits? While the `.svg` is of course editable, it will be a lot easier to work with the original file and create a new export.

robdodson commented:
> The source .excalidraw file is not being used, but the exported-from-it .svg is. Is there a better place to keep the source around in case we need to make edits? While the .svg is of course editable, it will be a lot easier to work with the original file and create a new export.

Maybe in a public Drive folder with a comment that contains a link to it?

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,1088 @@
+Teeing also is important because streams cannot be rewound or restarted, more about this later.
+
+<figure class="w-figure">
+  <img src="./streams.svg" width="2099" height="1127" alt="Diagram of a pipe chain consisting of a readable stream coming from a call to the fetch API that is then piped through a transform stream whose output is teed and then sent to the browser for the first resulting readable stream and to the service worker cache for the second resulting readable stream.">
```
```suggestion
  <!-- Original source file located at https://drive.google.com/file/d/17apgoyo6E5RAA_nwwM5Xg4FGiMr8y3mk/view?usp=sharing -->
    <img src="./streams.svg" width="2099" height="1127" alt="Diagram of a pipe chain consisting of a readable stream coming from a call to the fetch API that is then piped through a transform stream whose output is teed and then sent to the browser for the first resulting readable stream and to the service worker cache for the second resulting readable stream.">
```

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,1089 @@
+  network and process them as desired.
+authors:
+  - thomassteiner
+date: 2021-02-03
```
```suggestion
date: 2021-02-17
```

tomayac commented:
> > The source .excalidraw file is not being used, but the exported-from-it .svg is. Is there a better place to keep the source around in case we need to make edits? While the .svg is of course editable, it will be a lot easier to work with the original file and create a new export.
> 
> Maybe in a public Drive folder with a comment that contains a link to it?

Done. Deleted the file from the repo and added a Drive link to the Markdown text.

jpmedley commented:
@tomayac Is this ready for me?

tomayac commented:
@kaycebasques This was approved by @jpmedley, but I can't merge yet. It looks like someone else needs to override the "requested changes" blocker (the changes were addressed). Or alternatively, @jakearchibald and/or @beaufortfrancois, could you approve, too?

tomayac commented on src/site/content/en/blog/streams/index.md:
```diff
@@ -0,0 +1,1089 @@
+  network and process them as desired.
+authors:
+  - thomassteiner
+date: 2021-02-17
```
```suggestion
date: 2021-02-18
```

kaycebasques commented:
@tomayac I dismissed the review that was blocking you
